src/pagination.rs

```diff
@@ -242,6 +247,142 @@ fn sort_posts(posts: &mut Vec<&liquid::value::Value>, config: &PaginationConfig)
     })
 }
 
+// Categories
+
+#[derive(Debug)]
+struct Category<'a> {
+    cat_name: liquid::value::Array,
+    posts: Vec<&'a liquid::value::Value>,
+    sub_cats: Vec<Category<'a>>,
+}
+
+impl<'a> Category<'a> {
+    fn new(name: &liquid::value::Array) -> Self {
+        Category {
+            cat_name: name.to_vec(),
+            posts: vec![],
+            sub_cats: vec![],
+        }
+    }
+
+    fn add_post(&mut self, post: &'a liquid::value::Value) {
+        self.posts.push(&post)
+    }
+}
+
+// only used in parse_categories_list, to transform liquid::value::Array to type with Ord trait
+fn liquid_array_to_string(arr: &liquid::value::Array) -> String {
+    let mut s: String = arr
+        .iter()
+        .map(|i| {
+            let mut s = i.to_str().to_string();
+            s.push('/');
+            s
+        })
+        .collect();
+    s.pop(); // remove last '/'
+    s
+}
+
+// construct a hierarchy of Categories with their posts from a list of categories
+fn parse_categories_list<'a, 'b>(
+    parent: &'b mut Category<'a>,
+    cur_idx: usize,
+    categories: &[liquid::value::Value],
+    post: &'a liquid::value::Value,
+) -> Result<()> {
+    if cur_idx <= categories.len() {
+        let cat_full_name = categories[..cur_idx].to_vec();
+        let mut cur_cat = if let Ok(idx) = parent
+            .sub_cats
+            .binary_search_by_key(&liquid_array_to_string(&cat_full_name), |c| {
```

Comment:
imo it'd be better to use `binary_search_by` and provide our own comparator than doing a lot of allocations in key-generation.

Reply:
I'm confused. Here is my comparison function:
```rust
fn compare_category_path(p1: &liquid::value::Array, p2: &liquid::value::Array) -> Ordering {
    p1.iter()
        .partial_cmp(p2.iter())
        .expect("Arrays of same hierarchy level should be fully comparable")
}
```

I'm using it in a sort:
```rust
parent.sub_cats
          .sort_by(|c1, c2| compare_category_path(&c1.cat_path, &c2.cat_path));
```

And in the `binary_search_by`:
```rust
parent
            .sub_cats
            .binary_search_by(|c| 
                compare_category_path(&cat_full_path, &c.cat_path))
```

It does not work.

But if I change the order of the parameters in `binary_search_by`, it works…
How so? What am I missing?

Reply:
So `compare_category_path(&cat_full_path, &c.cat_path))` fails but `compare_category_path(&c.cat_path, &cat_full_path))` succeeds?

Reply:
Yes, which bother me a lot ^^'

Reply:
By looking at https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by
It seems that the `seek` is on second position as well. 
So… this is normal?